### PR TITLE
[dtensor] support convolution ops

### PR DESCRIPTION
**Pull request by KingsleyLiu-NV**
This PR creates a prototype of training convolutional neural networks based on DTensor.

- Register required ops and implement operator dispatch
- Add unit tests and example

Basically, we shard the activations and replicate the model weights in this prototype. We can scale out to multiple GPUs and reduce the per-GPU memory footprint with this approach, and achieve weak scaling in terms of training performance (i.e., time per iteration). 

Reference log (on 2xA100 GPU):

Unit Test
```bash
root@luna-prod-78-80gb:/pytorch# python3 test/distributed/_tensor/test_convolution_ops.py
/opt/conda/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: 0TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (Triggered internally at /opt/conda/conda-bld/pytorch_1699257304556/work/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2170.)
  return F.conv2d(input, weight, bias, self.stride,
/opt/conda/lib/python3.10/site-packages/torch/nn/modules/conv.py:456: UserWarning: 0TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (Triggered internally at /opt/conda/conda-bld/pytorch_1699257304556/work/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:2170.)
  return F.conv2d(input, weight, bias, self.stride,
..
----------------------------------------------------------------------
Ran 2 tests in 30.354s

OK
root@luna-prod-78-80gb:/pytorch# python3 test/distributed/_tensor/test_other_ops.py
[rank0]:[W ProcessGroupNCCL.cpp:2170] Warning: 0TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (function operator())
[rank0]:[W ProcessGroupNCCL.cpp:2170] Warning: 0TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (function operator())
[rank1]:[W ProcessGroupNCCL.cpp:2170] Warning: 0TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (function operator())
[rank1]:[W ProcessGroupNCCL.cpp:2170] Warning: 0TORCH_NCCL_AVOID_RECORD_STREAMS=1 has no effect for point-to-point collectives. (function operator())
...
----------------------------------------------------------------------
Ran 3 tests in 16.343s

OK
```
ConvNeXt Example
```bash
root@luna-prod-78-80gb:/pytorch# python3 torch/distributed/_tensor/examples/convnext_example.py
rank 3, 20 iterations, latency     584.80 ms, forward     102.84 ms, backward     297.80 ms, max reserved    16.34 GiB, max allocated    14.75 GiB
rank 1, 20 iterations, latency     584.64 ms, forward     104.85 ms, backward     297.60 ms, max reserved    16.40 GiB, max allocated    14.74 GiB
rank 0, 20 iterations, latency     584.48 ms, forward     104.64 ms, backward     297.90 ms, max reserved    16.39 GiB, max allocated    14.75 GiB
rank 2, 20 iterations, latency     584.96 ms, forward      93.21 ms, backward     297.95 ms, max reserved    16.40 GiB, max allocated    14.74 GiB
```

@wanchaol @fduwjj FYI
 

cc @wanchaol @XilunWu

### CODE REVIEW

File: torch/distributed/_tensor/tp_conv.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    stride, padding, dilation = local_tensor_args[3:6]
+
+    assert _is_supported(in_tensor.shape, weight.shape, stride, padding, dilation)
+
+    if not _requires_data_exchange(padding):
+        local_results = op_call(*local_tensor_args, **local_tensor_kwargs)
+        return local_results
+    else:
+        # step 0 compute the overlap pixels of the input tensor
+        d = weight.shape[3] - 1
```
Comment (wanchaol) [inline torch/distributed/_tensor/tp_conv.py:59]: I am trying to see if we should make these batch send/recv calls be a type of redistribute/resharding, so instead of having a custom convolution op implemented, and special case in torch_dispatch, we should try to see if:
- for the case of input data exchange, we can make this input send/recvs be a data order permutation (i.e. sth like mesh [0, 1, 2, 3] to mesh [1, 2, 3, 0]).
- I see duplicate comms happen in the backward op too, we should merge these two into a common function.

File: torch/distributed/_tensor/tp_conv.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        d1 = d // 2
+        d2 = d - d1
+        assert d1 + d2 == d
+        right = (rank + 1) % size
+        left = (rank - 1 + size) % size
+
+        # step1 dist comms
+        send_to_right = in_tensor[:, :, :, -d1:].contiguous()
```
Comment (wanchaol) [inline torch/distributed/_tensor/tp_conv.py:67]: It's a bit hard to follow what this dist comm is doing, could you add some comments to explain what this batch send/recv is trying to achieve for the conv input?

File: torch/distributed/_tensor/ops/other_ops.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+
+@register_prop_rule(aten.bernoulli_.float)
+def bernoulli_rules(op_schema: OpSchema) -> OutputSharding:
+    input_spec = op_schema.args_schema[0]
+    return OutputSharding(input_spec)
+
+
+@register_prop_rule(aten.nll_loss_forward.default)
+def nll_loss_forward_rules(op_schema: OpSchema) -> OutputSharding:
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/other_ops.py:111]: Given that those "other ops" are not related to convolution specifically, it would be better if we can separate the other ops enablement with convolution itself with some tests

**Review by wanchaol — COMMENTED** [review]
Thanks for contributing! Convolution op is quite important and also hard to implement in the distributed tensor context, so glad that you already get this working :)

I have a few comments inlined with regard to how the ops (for conv and other ops) are implemented currently. The main thing is that we should see if convolution op can be implemented like other ops, without special casing to do pre/post communication.

File: torch/distributed/_tensor/tp_conv.py (hunk 2 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/tp_conv.py:67]: Sure, I have sent you the slides about implementation details in the slack channel.

File: torch/distributed/_tensor/ops/other_ops.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+import torch
+from torch.distributed._tensor.op_schema import OpSchema, OutputSharding
+from torch.distributed._tensor.ops.utils import register_prop_rule
+from torch.distributed._tensor.placement_types import DTensorSpec, TensorMeta
+
+aten = torch.ops.aten
+
+
+@register_prop_rule(aten.slice_backward.default)
+def slice_backward_rules(op_schema: OpSchema) -> OutputSharding:
```
[the remark this answers is not in the document]
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/ops/other_ops.py:14]: I prefer to keep it the current way and  refactor it later.

File: torch/distributed/_tensor/tp_conv.py (hunk 1 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/tp_conv.py:125]: I have merged communications as a function for conv fwd/bwd.

Regarding replacing batch send/recv calls with redistribute/resharding, we can keep the discussion, and I think it will not happen in this PR.

File: torch/distributed/_tensor/ops/other_ops.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        dim_map,
+        sums,
+        tensor_meta=grad_input_tensor_meta,
+    )
+
+    return OutputSharding(grad_input_spec)
+
+
+@register_prop_rule(aten.native_layer_norm.default)
+def layer_norm_rules(op_schema: OpSchema) -> OutputSharding:
```
[the remark this answers is not in the document]
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/ops/other_ops.py]: Thanks for letting me konw, I have removed it accordingly.

File: torch/distributed/_tensor/ops/other_ops.py (hunk 3 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/ops/other_ops.py:47]: Can you elaborate more on this? Currently we need to register `slice_backward`, `bernoulli_`, `nll_loss_forward` and `nll_loss_backward` within [other_ops.py](https://github.com/KingsleyLiu-NV/pytorch/blob/fea_dtensor_conv/torch/distributed/_tensor/ops/other_ops.py) to run the tp training example [convnext_example.py](https://github.com/KingsleyLiu-NV/pytorch/blob/fea_dtensor_conv/torch/distributed/_tensor/examples/convnext_example.py).

Is it good to keep other_ops.py as it is, and add some unit tests by creating a new file `test_other_ops.py` within [test/distributed/_tensor](https://github.com/KingsleyLiu-NV/pytorch/tree/fea_dtensor_conv/test/distributed/_tensor)?

File: test/distributed/_tensor/test_convolution_ops.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            distribute_tensor(param, device_mesh, dist_spec)
+        )
+        name = "_".join(name.split("."))
+        module.register_parameter(name, dist_param)
+
+
+class DistConvolutionOpsTest(DTensorTestBase):
+    @property
+    def world_size(self) -> int:
+        # hard code world size to 2
```
Comment (XilunWu) [inline test/distributed/_tensor/test_convolution_ops.py:42]: shall we change the world size to 4 to align with other tests?

File: test/distributed/_tensor/test_convolution_ops.py [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        return 2
```
Comment (XilunWu) [inline test/distributed/_tensor/test_convolution_ops.py:43]: shall we change the world size to 4 to align with other tests?

File: torch/distributed/_tensor/dispatch.py [hunk 8]
```diff
@@ -120,6 +122,14 @@ def operator_dispatch(
...
     sharding_propagator: ShardingPropagator,
 ) -> object:
+    # Redistribute grad_output tensor to the same placement as input tensor
+    args = list(args)
+    if op_call == torch.ops.aten.convolution_backward.default:
+        assert isinstance(args[0], dtensor.DTensor) and isinstance(
+            args[1], dtensor.DTensor
+        )
+        args[0] = args[0].redistribute(args[1].device_mesh, args[1].placements)
+    args = tuple(args)
```
Comment (XilunWu) [inline torch/distributed/_tensor/dispatch.py:132]: @wanchaol shouldn't this resharding happen in `_operator_dispatch`?

File: torch/distributed/_tensor/dispatch.py [hunk 9]
```diff
@@ -293,6 +303,12 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
                 cast(DTensorSpec, args_schema[0])
             ):
                 local_results = op_call(*local_tensor_args, **local_kwargs)
+        elif op_call == torch.ops.aten.convolution.default:
+            local_results = tp_convolution(op_call, *local_tensor_args, **local_kwargs)
+        elif op_call == torch.ops.aten.convolution_backward.default:
+            local_results = tp_convolution_backward(
+                op_call, *local_tensor_args, **local_kwargs
+            )
```
Comment (XilunWu) [inline torch/distributed/_tensor/dispatch.py:311]: @wanchaol what can we do to avoid this customized DTensor op?

File: torch/distributed/_tensor/examples/convnext_example.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+        self.data_format = data_format
+        if self.data_format not in [torch.contiguous_format, torch.channels_last]:
+            raise NotImplementedError
+        self.normalized_shape = (normalized_shape,)
+
+    def forward(self, x):
+        if self.data_format == torch.channels_last:
+            return F.layer_norm(
+                x, self.normalized_shape, self.weight, self.bias, self.eps
+            )
```
Comment (XilunWu) [inline torch/distributed/_tensor/examples/convnext_example.py:39]: note: I'm writing DTensor layer_norm_backward op but so far we only have the forward one.

File: torch/distributed/_tensor/tp_conv.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    assert len(local_tensor_args) == 9
+
+    rank = dist.get_rank()
+    size = dist.get_world_size()
+    in_tensor = local_tensor_args[0]
+    weight = local_tensor_args[1]
+    bias = local_tensor_args[2]
+    stride, padding, dilation = local_tensor_args[3:6]
+
+    assert _is_supported(in_tensor.shape, weight.shape, stride, padding, dilation)
```
Comment (XilunWu) [inline torch/distributed/_tensor/tp_conv.py:118]: need to fix linter complaints ;-)

**Review by XilunWu — COMMENTED** [review]
Thanks for the hard work! I'm working on backward layer norm support, wish it's not blocking you right now. Left some questions.

File: torch/distributed/_tensor/ops/other_ops.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/ops/other_ops.py]: Will this PR need `layer_norm_backward`?

[the remark this answers is not in the document]
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/ops/other_ops.py]: Yes, it need it to run the example: [convnext_example.py](https://github.com/KingsleyLiu-NV/pytorch/blob/fea_dtensor_conv/torch/distributed/_tensor/examples/convnext_example.py#L154)

I have verified that it works after removing my registration in [other_ops.py](https://github.com/pytorch/pytorch/pull/113123/files#diff-92d32796b3f857e1d7b6ca8ad08f06edaa46945e9908a2b25c14140dd4b7d533)

File: test/distributed/_tensor/test_convolution_ops.py (hunk 7 above)
Comment (KingsleyLiu-NV) [reply test/distributed/_tensor/test_convolution_ops.py:43]: Absolutely we can.

File: torch/distributed/_tensor/tp_conv.py (hunk 11 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/tp_conv.py:118]: I have fixed some linter errors with `lintrunner f xxx.py`. But there are still errors about which I am not sure. For example:

```bash
  Error (MYPY) [union-attr]
    Item "None" of "Optional[Tensor]" has no attribute "data"

        111  |def init_weights(m):
        112  |    if type(m) == nn.Conv2d or type(m) == nn.Linear or type(m) == nn.LayerNorm:
        113  |        m.weight.data.fill(1.0)
    >>> 114  |        m.bias.data.fill_(0.0)
        115  |
        116  |
        117  |class ConvNeXt(nn.Module):
```

I have create a document listing them. It will be appreciated if you can give me some hint :)
[linter_errors.docx](https://github.com/pytorch/pytorch/files/13314308/linter_errors.docx)

File: torch/distributed/_tensor/examples/convnext_example.py (hunk 10 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/examples/convnext_example.py:39]: I see. `F.layer_norm` and `nn.LayerNorm` are not actually used in this example.

File: test/distributed/_tensor/test_convolution_ops.py (hunk 7 above)
Comment (KingsleyLiu-NV) [reply test/distributed/_tensor/test_convolution_ops.py:43]: Done.

File: torch/distributed/_tensor/ops/other_ops.py (hunk 3 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/ops/other_ops.py:50]: I have add unit tests for other ops and showcase the reference log in the description in this PR.

File: torch/distributed/_tensor/examples/convnext_example.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+
+        forward_time += t2 - t1
+        backward_time += t4 - t3
+    torch.cuda.synchronize(device)
+    end = time.time()
+    max_reserved_memory = torch.cuda.max_memory_reserved(device)
+    max_allocated_memory = torch.cuda.max_memory_allocated(device)
+    print(
+        f"rank {rank}, {ITER_TIME} iterations, latency {(end - start)/ITER_TIME*1000:10.2f} ms, forward {forward_time/ITER_TIME*1000:10.2f} ms, backward {backward_time/ITER_TIME*1000:10.2f} ms, max reserved {max_reserved_memory/1024/1024/1024:8.2f} GiB, max allocated {max_allocated_memory/1024/1024/1024:8.2f} GiB"
+    )
```
Comment (XilunWu) [inline torch/distributed/_tensor/examples/convnext_example.py:245]: split this long line into multiple lines, for example:
```
print((
    line1
    line2
    ...
 ))
```

File: test/distributed/_tensor/test_convolution_ops.py (hunk 7 above)
Comment (XilunWu) [reply test/distributed/_tensor/test_convolution_ops.py:43]: Thx for addressing the comment!

File: torch/distributed/_tensor/examples/convnext_example.py (hunk 12 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/examples/convnext_example.py]: Fixed.

File: torch/distributed/_tensor/dispatch.py [hunk 13]
```diff
@@ -88,6 +91,14 @@ def dispatch(
         """
         Main dispatching logic
         """
+        # Redistribute grad_output tensor to the same placement as input tensor
+        args = list(args)
```
Comment (wanchaol) [inline torch/distributed/_tensor/dispatch.py:95]: hmmm I think we shouldn't do those checks and run redistribute, this is really hacky. I would prefer if this thing can be implemented as  sharding propagation + resharding workflow. If you want to those custom hacks instead of supporting sharding propagation directly. we should not hijack to op dispatching logic like this.

So if you want to keep the current customization, you can probably do it in this way:
1. register `aten.convolution_backward.default` and `aten.convolution.default` to the custom_op_handler in the `OpDispatcher` https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/dispatch.py#L78
2. the custom handler for these two conv ops should call into the `tp_conv.py`, where inside tp_conv.py, you can put these custom logic into the custom dispatch functions(i.e. tp_convolution, tp_convolution_backward), where you could do all custom things like unwrapping DTensor to get local shards, run conv rule propagation, run tp_conv, re-wrapping local outputs with DTensors.

File: torch/distributed/_tensor/ops/other_ops.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/other_ops.py:1]: Let's rename this `other_ops.py` to sth like `experimental_ops.py`

File: torch/distributed/_tensor/examples/convnext_example.py [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+import torch.nn as nn
+from torch.distributed._tensor import (
+    DeviceMesh,
+    distribute_module,
+    distribute_tensor,
+    Replicate,
+    Shard,
+)
+
+warnings.filterwarnings("ignore", category=UserWarning)
```
Comment (wanchaol) [inline torch/distributed/_tensor/examples/convnext_example.py:21]: why we need to filter out warning here?

File: torch/distributed/_tensor/examples/convnext_example.py [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+        f"rank {rank}, forward {forward_time/ITER_TIME*1000:10.2f} ms, backward {backward_time/ITER_TIME*1000:10.2f} ms"
+    )
+    print(
+        f"rank {rank}, max reserved {max_reserved/1024/1024/1024:8.2f} GiB, max allocated {max_allocated/1024/1024/1024:8.2f} GiB"
+    )
+    dist.destroy_process_group()
+
+
+if __name__ == "__main__":
+    mp.spawn(test_tp_convnext_train, args=(WORLD_SIZE,), nprocs=WORLD_SIZE, join=True)
```
Comment (wanchaol) [inline torch/distributed/_tensor/examples/convnext_example.py:260]: maybe we can just call `torchrun` without manually setting up multiprocess or the world process group, and just add comments about how to run this file with `torchrun`, but we can do this refactor in follow up PRs

**Review by wanchaol — COMMENTED** [review]
Thanks for fixing all the lint errors! Please see inlined comments.

I think the other changes looks good, my major comment now is that we shouldn't hijack into the dispatch logic like this in `dispatch.py`, given that we are doing custom op handing, let's just make convolution now a custom op handler, and handle the custom logic inside `tp_conv.py`

File: torch/distributed/_tensor/dispatch.py (hunk 13 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py:95]: @wanchaol It is the problem of `sharding propagation + resharding` of `torch.Tensor.mean`, which undermines the `Shard(3)` placement of activations of convolution.

image

If the sharding propagation of mean([-2,-1]) can be fixed properly, I think this hijack can be eliminated. The convolution per se does not require any custom hacks actually,

File: torch/distributed/_tensor/ops/other_ops.py (hunk 14 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/ops/other_ops.py]: Done

File: torch/distributed/_tensor/examples/convnext_example.py (hunk 15 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/examples/convnext_example.py]: Thanks for reminding. I have removed it.

File: torch/distributed/_tensor/examples/convnext_example.py (hunk 16 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/examples/convnext_example.py:258]: It can be refactored later, given that other examples also use `mp.spawn`：https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/examples/checkpoint_example.py#L182

File: torch/distributed/_tensor/dispatch.py (hunk 13 above)
Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:95]: Copy my comment over:

Thanks for the super detailed explaination! The shardings for the fwd and bwd for a specific operator does not have to be aligned as it is mostly determined by the backward formula as you observed, that's why we have the resharding logics to make sure if the backward ops are sharding propagated differently it would perform resharding when necessary, currently using OpStrategy we can specify a op only "accepts" Shard(3) so that if the argument is not Shard(3) it would perform argument resharding to Shard(3) first. That's why I was suggesting we should try to make the ops uses the Op Strategy approach. Additionally, we essentially dispatch convolution to tp_convolution and tp_convolution_bwd call instead of using the original aten.convolution for local shard computation, this is not a uniform way as other ops.

But if you want to land the currently way of doing sharding propagation for convolution ops without the resharding logic, I am fine too, but we need to avoid polluting the dispatch.py, I think you can register the custom_handler to custom handle both convolution and convolution backward

Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:95]: The tp_conv custom handler we can register should be relatively simple, after register a custom handler, you just need to do the following things in the handler:
1. unwrap arguments from DTensor to local_shard tensor and the sharding specs, this code can mostly copy from the dispatch.py
2. run conv and conv_backward sharding propagation rule to get the output sharding
3. run tp_conv and tp_conv_backward
4. rewrap the outputs to DTensor with the output sharding, this part can just call the OpDispatcher.wrap

Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py:95]: I see, and the first two steps seem to introduce duplicated code, is it good? Also, which place is better to put `aten_convolution_handler `and `aten_convolution_backward_handler `, `dispatch.py` or `tp_conv.py`?

Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py:95]: If switching to `OpStrategy` for convolution can make cleaner code, I prefer to go for this. Do you have some suggestions on required changes?

Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py:95]: I decided to keep using `register_prop_rule` and use custom handlers in this PR. Following the above 4 steps, I have this code snippet:

```
def convolution_backward_handler(
    op_call: torch._ops.OpOverload,
    args: Tuple[object, ...],
    kwargs: Dict[str, object],
) -> object:
    # Redistribute grad_output tensor to the same placement as input tensor
    args = list(args)
    if op_call == torch.ops.aten.convolution_backward.default:
        assert isinstance(args[0], dtensor.DTensor) and isinstance(
            args[1], dtensor.DTensor
        )
        args[0] = args[0].redistribute(args[1].device_mesh, args[1].placements)
    args = tuple(args)
    local_tensor_args = [
        arg.to_local() if isinstance(arg, dtensor.DTensor) else arg for arg in args
    ]
    local_tensor_args = cast(Tuple[object, ...], local_tensor_args)
    local_kwargs = kwargs
    local_results = tp_convolution_backward(op_call, local_tensor_args, local_kwargs)
  
    # HOW CAN I GET output_sharding WITHOUT DUPLICATING TOO MUCH CODE???

    return OpDispatcher.wrap(local_results, output_sharding.output_spec)
```

The remaining question is how to get  `output_sharding ` in this handler function without duplicating too much code. Any suggestions are appreciated.

Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py:95]: I have added customer handlers as suggested. Thank you so much for refactoring work to make components of op dispatch more usable.

File: torch/distributed/_tensor/dispatch.py (hunk 8 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py]: I have moved it to `convolution_backward_handler` to avoid this pollution.

File: torch/distributed/_tensor/dispatch.py (hunk 9 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py]: I have moved it to `convolution_backward_handler` to avoid this pollution.

File: torch/distributed/_tensor/dispatch.py [hunk 17]
```diff
@@ -55,6 +56,64 @@ def is_same_size_handler(
...
+    # extract local tensor and sharding infos to a OpInfo
+    op_info = dtensor.DTensor._op_dispatcher.unwrap_to_op_info(op_call, args, kwargs)
+
+    # sharding propagation
+    dtensor.DTensor._op_dispatcher.sharding_propagator.propagate(op_info)
+    output_sharding = op_info.output_sharding
+    assert output_sharding is not None, "output sharding should not be None"
+
+    # local propagation
+    local_tensor_args = [
```
Comment (wanchaol) [inline torch/distributed/_tensor/dispatch.py:73]: local_tensor_args can be accessed directly from `op_info`, i.e. op_info.local_args.

File: torch/distributed/_tensor/dispatch.py [hunk 18]
```diff
@@ -55,6 +56,64 @@ def is_same_size_handler(
     return lhs.shape == rhs.shape
 
 
+def convolution_handler(
```
Comment (wanchaol) [inline torch/distributed/_tensor/dispatch.py:59]: nit: maybe we can put these two handlers in the `tp_conv.py` and import the handlers directly?

File: torch/distributed/_tensor/dispatch.py [hunk 19]
```diff
@@ -55,6 +56,64 @@ def is_same_size_handler(
...
+    # extract local tensor and sharding infos to a OpInfo
+    op_info = dtensor.DTensor._op_dispatcher.unwrap_to_op_info(op_call, args, kwargs)
+
+    # sharding propagation
+    dtensor.DTensor._op_dispatcher.sharding_propagator.propagate(op_info)
+    output_sharding = op_info.output_sharding
+    assert output_sharding is not None, "output sharding should not be None"
+
+    # local propagation
+    local_tensor_args = [
```
Comment (wanchaol) [inline torch/distributed/_tensor/dispatch.py:104]: ditto: iiuc you don't need to manually unwrap the local tensor args, as they already been unwrapped as part of `unwrap_to_op_info`

**Review by wanchaol — APPROVED** [review]
lgtm, have some small suggestions inlined, please address them before landing. thanks for contributing!

File: torch/distributed/_tensor/dispatch.py (hunk 19 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py:104]: Done.

File: torch/distributed/_tensor/dispatch.py (hunk 18 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py]: Sure.

File: torch/distributed/_tensor/dispatch.py (hunk 17 above)
Comment (KingsleyLiu-NV) [reply torch/distributed/_tensor/dispatch.py]: Thanks for letting me know. I have done it.
